README.md
# BP_PROJ


### piskvorky_MP [➥](piskvorky_MP/)
- Piškvorky pro až 5 hráčů
- 1x arduino Due jako server, max 5x arduino Ethernet s dotykovým displejem jako client
- připojovánou přes router s DHCP (serveru nutno přidělit pevnou IP)
- aktualizovaný návod na ovládání [zde](https://github.com/janzavorka/BP_PROJ/blob/master/piskvorky_MP/manual.md)
---

### Piskvorky_red [➥](Piskvorky_red/)
- Původní verze verze hry pouze pro dva hráče
- HW: 2x Arduino Ethernet, 2x 2,4" dotykový display
- jedno arduino jako server, druhé jako client (pevná IP adresa)
---

### Knihovny [➥](Knihovny/)
- Knihovny použité pro ovládání (nemusí být aktuální verze)
- Domovské stránky knihoven:
    - [SimpleTimer](https://github.com/schinken/SimpleTimer)
    - Pro ovládání displeje:
        - [MCUFRIEND_kbv](https://github.com/prenticedavid/MCUFRIEND_kbv)
        - [Adafruit_TouchScreen](https://github.com/adafruit/Adafruit_TouchScreen)
---

### Menu [➥](Menu/)
- Návrh původního menu pro piškvorky
---

### piskvorky_case [➥](piskvorky_case/)
- Návrh krabiček pro jednotlivá zařízení ([Autodesk Inventor Professional 2019 Student Edition](https://www.autodesk.cz/products/inventor/overview))
- Realizace pomocí 3D tisku

| <img src="piskvorky_case/client_ArdEthernet_case/Sestava_krabicka2.png" width="49%" alt="Krabička pro clienta" /> <img src="piskvorky_case/client_ArdEthernet_case/Sestava_krabicka1.png" width="49%" alt="Krabička pro clienta" /> |
|:---------:|
|Krabička pro clienta (Arduino Ethernet s 2,4" displejem)|

| <img src="piskvorky_case/server_ArdDue_case/sestava_krabicka2.png" width="49%" alt="Krabička pro server" /> <img src="piskvorky_case/server_ArdDue_case/sestava_krabicka1.png" width="49%" alt="Krabička pro server" /> |
|:---------:|
|Krabička pro server (Arduino Due s Ethernet modulem + tlačítka a signalizační LED)|

 ---

### piskvorky_MP_old [➥](piskvorky_MP_old/)
- Starší funkční verze projektu
---

### BAP [➥](BAP/zaverecna_prace_FEL/)
-  Bakalářská práce, předmět BBAP15	 [(pdf)](BAP/zaverecna_prace_FEL/main.pdf)
---

### Zprava_BPROJ [➥](Zprava_BPROJ/)
-  Výstupní zpráva z celého projektu pro předmět B2BPROJ4 [(pdf)](Zprava_BPROJ/BPROJ_zprava.pdf)
---

### RGB_LED_cal [➥](RGB_LED_cal/)
- Slouží ke kalibraci barev pro RGB LED u serveru (Arduino Due)
- RGB LED připojení:
    - R -> D6
    - G -> D7
    - B -> D8
- RGB LED se připojí na příslušné PWM piny, změna barev a celkové intenzity probíhá pomocí sériového monitoru
- Příkaz _"help"_ vypíše nápovědu přímo na sériový monitor
- Příkaz: _"led rrr,ggg,bbb,inten"_, kde:
    - _rrr,ggg,bbb_ jsou voleny z <0;&nbsp;255> a reprezentují jednotlivé barvy
    - _inten_ je volena z <0;&nbsp;100> a reprezentuje celkovou intenzitu LED
 ---

 ### Notes.md [➥](Notes.md)
 - Několik zjištění a postřehů z tvorby tohoto projektu


piskvorky_MP/CHANGELOG.md
# BP_PROJ: Piskvorky MP

### 10.05.2019
- Změna komunikačního portu na 55555, dle RFC6335
- Úprava komnetřá v kódu pro klienta

### 06.05.2019
- Úprava MAC adresa na ty, které mají desky Arduino Ethernet přiřazené od výrobce (na štítku zespodu)
- Výměna displeje pro klienta číslo 3, přidání kalibračních dat pro nový displej

### 05.05.2019
- Server:
  - Přidáno nastavení síťového režimu (_ETHMODE\_DHCP_ a _ETHMODE\_STATIC_, stejný princip jako u clienta)
  - Upraveny podmínky u direktiv _#elif_ (přidáno _defined()_)
- Client:
    - Upraveny podmínky u direktiv _#elif_ (přidáno _defined()_)


### 21.04.2019
- Program pro clienta:
  - Upraveny přidání schémata v nastavení pro jednotlivé clienty pomocí direktiv (#define CLIENT_x_)
  - Přidán režim doutykové plochy displeje _TOUCH\_LANDSCAPE_ případně _TOUCH\_PORTRAIT_
  - Přidán síťové režimy:
    - _ETHMODE\_DHCP_: client získává IP adresu z DHCP serveru
    - _ETHMODE\_STATIC_: client používá statickou IP adresu nastavenou v kódu
  - Přidány komentáře ke konfiguracím

### 14.04.2019
- Program pro server:
  - Přidány možnost pro debuging
- Úprava manuálu pro piškvorky (_manual.md_): do tabulky přidán příkaz pro přepnutí na dalšího hráče a přidán popis funkce tlačítek

### 13.04.2019
- Program pro server:
  - Přidání příkazu pro odpojení hráčem
  - Přidána varianta, že pokud má server v kódu nastavenou adresu 0.0.0.0, nechá si IP adresu přiřadit DHCP serverem
  - Přídán příkaz pro přeskočení hráčem
  - Pokud běží hra a je stisknuto zelené tlačítko, hra se přesune na dalšího hráče
- Přidání hlavičky do zdrojových kódů, celkové úprava zdrojáků a doplnění komentářů
- Úprava readme.md: přidání odkazů (šipky) de jednotlivých složek
- Úprava manuálu pro piškvorky (_manual.md_): do tabulky přidán příkaz pro odpojení hráče
- Přesun projektu do složky _piskvorky_MP_ a přejmenování příslošných složek pro zdrojáky, úprava názvů v _readme.md_

### 09.03.2019
- Program pro server:
  - Doladěno funguvání indikační LED
  - Provedeno několik testů

- Program pro server:
  - Upravena funkce pro odesílání dat serveru
  - Přidána funkce pro kontrolu spojení se serverem (v případě, že došlo ke ztrátě spojení, je spojení zrušeno i ze strany clienta a vypíše se chybová hláška)
  - Přidáno vyhodnoce kódu devět (vyžádané odpojení ze strany serveru)
  - Kalibrace displejů
  - Drobné kosmetické úpravy kódu a doplnění komentářů
- Zhotoveny kalibrace Clientů 1 a 2 dle zjištění popsané v [notes](https://github.com/janzavorka/BP_PROJ/blob/master/Notes.md).


### 09.03.2019
- Dokončení programu pro clienta (kvůli místu v paměti odstraněna sériová linka)
- Malé úpravy v kódu pro server, odstraněna chyba v přijímací funkci

### 06.03.2019
- Program pro clienta: vylepšení vykreslování dat na displeji, začátek úprav vyhodnocování stisku tlačítek

### 03.03.2019
- Úprava programu pro clienta pro komunikaci se serverem

### 26.02.2019
- Automatické detekce ethernet kontroléru při zapnutí
- Pro kontroléry W5200 a W5500 kontrola připojení kabelu, chyba indikována blikající červenou LED
- Vylepšení a dodělání fukcí pro řízení hry
- Přidány funkce pro indikaci stavu hry pomocí RGB LED

### 25.02.2019
- Úprava funkcí pro řízení hry

### 23.02.2019
- Celkové přepracování programu pro server
- Pro server přidána LED dioda indikující stav
- Přidání zabezpečovacích funkcí pro komunikaci client->server a server -> client

### 12.12.2018
- Program pro clienta
 - pro všechny client desky zkalibrovány displeje, u clienta 3 nutná úprava v loopu (dislej má jinou orientaci)
 - oprava zobrazování hlášek (výhra, prohra, remíza)

### 09.12.2018
- Vytvoření desky s dvěma tlačítkama pro ovládání serveru
- Program pro server
 - přidání fce dvou tlačítek (start, stop), vyhodnocení stisku v loopu (nelze využít interrupty (rozhodilo by odesílání dat))

### 08.12.2018
- Použití Arduino Due jako server
- Program pro server
 - herní deska se odesílá po 8 bajtech
 - oprava funkce pro získání čísla dalšího hráče (docházelo k zacyklení)
 - přidána funkce pro kontrolu výhry
- Program pro clienta
 - Oprava zobrazování barevných koleček (nyní se zobrazují správně)
 - oprava zamrznutí hráče, kdy během jeho tahu dojde k resetu hry ze strany serveru

### 03.12.2018
- Program pro server
 - úprava funkce pro získání dalšího clienta

### 27.11.2018
- Program pro clienta
 - oprava připojovací funkce
 - doladění úvodních obrazovek


### 25.11.2018
- Práce na programu pro clienta:
 - Dodělány jednotlivé fáze hry
 - Funkce pro spojení se serverem
 - Mnoho dalších drobných úprav
- Program pro server
 - správné vyhodnocení odpojených clientů
 - při tahu daného clienta se pošle jen index dané pozice, ne celý herní board


### 12.11.2018
- Program pro clienta:
 - přidání třídy pro tlačítka

### 11.11.2018
- Příprava programu pro clienta
 - Úprava parametrů pro síťové připojení
 - Příprava rozložení packetu pro komunikaci
- Základ programu pro server


BAP/zaverecna_prace_FEL/codes/server_config.cpp
/* !!! ---------- KONFIGURACE - nastaveni ethernetu ---------- !!! */
#define ETHMODE_STATIC //Varianty: ETHMODE_DHCP; ETHMODE_STATIC
byte mac[] = {
  0xDE, 0xAD, 0xBE, 0xEE, 0xFE, 0x00
};
IPAddress serverAddress(10,0,0,8);
unsigned int localPort = 3333;
/* !!! ---------- KONEC nastaveni ethernetu ---------- !!! */

/* !!! ---------- KONFIGURACE - nastaveni pinu a LED ---------- !!! */
//Piny tlacitka start a reset
#define startPIN 2
#define resetPIN 3
//
//Piny na kterych jsou pripojenr jednotlive barvy LED, musi podporovat PWM
#define LED_red 7
#define LED_green 5
#define LED_blue 6
//
//Nastaveni maximalniho jasu vsech LED, interval <0; 100>
const byte LED_br = 40;
/* !!! ---------- KONEC nastaveni pinu a LED ---------- !!! */

/* !!! ---------- KONFIGURACE - nastaveni barev hracu ---------- !!! */
#define PLAYER1COLOR RED            //0xF800
#define PLAYER2COLOR GREEN          //0x07E0
#define PLAYER3COLOR PURPLE         //0x780F
#define PLAYER4COLOR GREENYELLOW    //0xAFE5
#define PLAYER5COLOR OLIVE          //0x7BE0
/* !!! ---------- KONEC nastaveni barev hracu ---------- !!! */


BAP/zaverecna_prace_FEL/codes/client_config.cpp
/* ---------- KONFIGURACE - nastaveni schemat----------*/
#define CLIENT1 //Vyber profilu pro klienta
/* ---------- KONEC - nastaveni schemat----------*/

/* ---------- KONFIGURACE - nastaveni mac adres----------*/
#ifdef CLIENT1
//Klient 1
byte mac[] = {
  0x90, 0xA2, 0xDA, 0x11, 0x08, 0x18
};
#elif defined(CLIENT2)
//Klient 2
byte mac[] = {
  0x90, 0xA2, 0xDA, 0x11, 0x09, 0x78
};
#elif defined(CLIENT3)
//Klient 3
byte mac[] = {
  0x90, 0xA2, 0xDA, 0x11, 0x08, 0xA0
};
#elif defined(CLIENT4)
//Klient 4
byte mac[] = {
  0xDE, 0xAD, 0xBE, 0xEE, 0xFE, 0xBD
};
#elif defined(CLIENT5)
//Klient 5
byte mac[] = {
  0xDE, 0xAD, 0xBE, 0xEE, 0xFE, 0xAD
};
#else
//Defaultni hodnoty
#warning "Je pouzita defaultni MAC adresa"
byte mac [] = {
  0xDE, 0xAD, 0xBE, 0xEE, 0xFE, 0xAB
}
#endif
/* ---------- KONEC - nastaveni mac adres----------*/

/* ---------- KONFIGURACE - nastaveni site ----------*/
#define ETHMODE_STATIC //Varianty: ETHMODE_DHCP; ETHMODE_STATIC

#ifdef ETHMODE_STATIC
  #ifdef CLIENT1
  IPAddress clientAddress(10,0,0,138);

  #elif defined(CLIENT2)
  IPAddress clientAddress(10,0,0,139);

  #elif defined(CLIENT3)
  IPAddress clientAddress(10,0,0,140);

  #elif defined(CLIENT4)
  IPAddress clientAddress(10,0,0,141);

  #elif defined(CLIENT5)
  IPAddress clientAddress(10,0,0,142);
  #else
  //Defaultni nastaveni
  IPAddress clientAddress(10,0,0,100);
  #warning "Pozor - je pouzito defaultni nastaveni IP, pro kazdeho clienta nutno zmenit"
  #endif
#endif
IPAddress serverAddress(10,0,0,8);  //Nastaveni IP adresy serveru
unsigned int localPort = 3333;      //Port
/* ---------- KONEC - nastaveni site ----------*/

/* ---------- KONFIGURACE - kalibrace displeje ----------*/
#ifdef CLIENT1
//Klient 1
#define TOUCH_XMIN 221
#define TOUCH_XMAX 950
#define TOUCH_YMIN 200
#define TOUCH_YMAX 950
#define TOUCH_LANDSCAPE
#elif defined(CLIENT2)
//Klient 2
#define TOUCH_XMIN  233
#define TOUCH_XMAX  937
#define TOUCH_YMIN  210
#define TOUCH_YMAX  910
#define TOUCH_LANDSCAPE
#elif defined(CLIENT3)
//Klient 3
#define TOUCH_XMIN 230
#define TOUCH_XMAX 960
#define TOUCH_YMIN 220
#define TOUCH_YMAX 920
#define TOUCH_LANDSCAPE
#elif defined(CLIENT4)
//Klient 4
#define TOUCH_XMIN 0
#define TOUCH_XMAX 100
#define TOUCH_YMIN 0
#define TOUCH_YMAX 100
#elif defined(CLIENT5)
//Klient 5
#define TOUCH_XMIN 0
#define TOUCH_XMAX 100
#define TOUCH_YMIN 0
#define TOUCH_YMAX 100
#else
//Defaultni nastaveni, je nutne upravit
#define TOUCH_XMIN 0
#define TOUCH_XMAX 100
#define TOUCH_YMIN 0
#define TOUCH_YMAX 100
#endif
/* ---------- KONEC - kalibrace displeje ----------*/
